packages/mds-compliance-engine/engine/count_processors.ts

```diff
@@ -0,0 +1,101 @@
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+import { Device, Geography, Policy, VehicleEvent, UUID, CountRule, Telemetry } from '@mds-core/mds-types'
+
+import { pointInShape, getPolygon, isInStatesOrEvents, isDefined } from '@mds-core/mds-utils'
+import { ComplianceResult } from '../@types'
+import { annotateVehicleMap, isInVehicleTypes, isPolicyActive, isRuleActive } from './helpers'
+
+export function isCountRuleMatch(
+  rule: CountRule,
+  geographies: Geography[],
+  device: Device,
+  // We throw out events that have no telemetry, so events are guaranteed
+  // to have telemetry.
+  event: VehicleEvent & { telemetry: Telemetry }
+) {
+  if (isRuleActive(rule)) {
+    for (const geography of rule.geographies) {
+      if (isInStatesOrEvents(rule, event) && isInVehicleTypes(rule, device)) {
+        const poly = getPolygon(geographies, geography)
+        if (poly && pointInShape(event.telemetry.gps, poly)) {
+          return true
+        }
+      }
+    }
+  }
+  return false
+}
+
+export function processCountPolicy(
+  policy: Policy,
+  events: (VehicleEvent & { telemetry: Telemetry })[],
+  geographies: Geography[],
+  devicesToCheck: { [d: string]: Device }
+): ComplianceResult | undefined {
+  const matchedVehicles: { [d: string]: { device: Device; rule_applied: UUID } } = {}
+  const overflowedVehicles: { [d: string]: { device: Device } } = {}
+  let countMinimumViolations = 0
+  let excess_vehicles_count = 0
+  if (isPolicyActive(policy)) {
+    const sortedEvents = events.sort((e_1, e_2) => {
+      return e_1.timestamp - e_2.timestamp
+    })
+    policy.rules.forEach(rule => {
+      const maximum = isDefined(rule.maximum) ? rule.maximum : Number.POSITIVE_INFINITY
+      let i = 0
+      sortedEvents.forEach(event => {
+        if (devicesToCheck[event.device_id]) {
+          const device = devicesToCheck[event.device_id]
+          if (isCountRuleMatch(rule as CountRule, geographies, device, event)) {
+            if (i < maximum) {
+              matchedVehicles[device.device_id] = { device, rule_applied: rule.rule_id }
+              /* eslint-reason need to remove matched vehicles */
+              /* eslint-disable-next-line no-param-reassign */
+              delete devicesToCheck[device.device_id]
+              delete overflowedVehicles[device.device_id]
+            } else {
+              overflowedVehicles[device.device_id] = { device }
+            }
+          }
+          /* If there's a match, increase i.
+           */
+          i += 1
+        }
+      })
```

Comment:
This should be a `reduce` where the accumulator in this case is your `i`

Reply:
Why? Seems like it's only a stylistic change, and kinda pointless when the increment doesn't always happen, and therefore the accumulator wouldn't always accumulate.

Reply:
That is precisely _what_ reduce statements exist for. In this case, it is not purely stylistic, because by making `i` mutable with a `let`, it can be changed anywhere else that it's in scope.

Reply:
All right, fair point about the mutability of `i`.

Reply:
But if you really want to get technical, the point of a `reduce` is to take a list and then return a single value. `i` isn't what you want to return and shouldn't be the accumulator if it's not the value that the whole thing needs to evaluate to.

Reply:
I don't really understand why `i` isn't what you want to return from the iteration, it's not like this is _returning_ anything at all currently. Why not make it closer to functional programming and actually have it return something? That is much cleaner than mutating a higher-scoped variable within a for loop. Sure, there are side-effects within this loop (perhaps something we can clean up), but making it a reducer which returns a value of interest is a step in the right direction.

Reply:
Because `i` isn't an interesting variable and the whole point is to return a list of vehicles that have been annotated somehow. A better candidate for an accumulator would probably be `matchedVehicles`. I can rewrite this with `i` as an indexed parameter, either for a `for` loop or a `reduce`.

Reply:
Actually, as I'm looking at this, is `i` incrementing on matches? Looks like it just increments upon every inner-most loop. I'd expect the increment of `i` to be a few lines above if it only incremented upon a match.

Reply:
> Why not make it closer to functional programming and actually have it return something? That is much cleaner than mutating a higher-scoped variable within a for loop. 

I think there are other kinds of cleanliness to consider here. Given the amount of state that has to be managed (`i`, `matchedVehicles`, and `overflowedVehicles`), it's IMO more readable to have them initialized in one place, outside of the loop. There's a very good chance that whoever has to touch this code next is not you or me, and I think my experience is going to be more representative of future maintainers' since you are the only one who was developing the overflow logic and writing the code at the same time. I even wrote a few `reduce` functions in the tests when I didn't feel they were necessary just to prove to myself that I wasn't a dim bulb who had a hard time reading or writing them. For _this_ particular piece of code, I'd like to keep my higher-scoped variables.

Reply:
I'm a little late here, but I will just add that personally, I try to avoid the use of `let`, `delete`, and other forms of mutating state in general. With there being a `forEach` in the mix above as well, this sure has all the signs of an operation that could be implemented via `reduce` ... there is nothing that says the accumulator has to be a single/scalar value; think of the accumulator as being the state object (with multiple properties) that is being managed.

Reply:
I do find `reduce` a little harder to read than `for` in many cases